exec.vu.tags documentation (#468)
grafana/k6-docs: src/data/markdown/docs/02 javascript api/06 k6-execution.md

| iterationInScenario | integer | The identifier of the iteration in the current scenario. |
| idInInstance        | integer | The identifier of the VU across the instance. |
| idInTest            | integer | The globally unique (across the whole test run) identifier of the VU. |

> | tags                | object  | It gives control over VU's Tags. The Tag will be included in every metric emitted by the VU and the Tag's state is maintained across different iterations until the VU exists. |

Reviewer: I've noticed this the previous time too but didn't get immediately what the problem is: all other items in the table start with a noun :sweat_smile: So to keep more in line, maybe like this:
```suggestion
| tags                | object  | The map that gives control over VU's Tags. The Tags will be included in every metric emitted by the VU and the Tags' state is maintained across different iterations until the VU exists. |
```

> | tags                | object  | The map that gives control over VU's Tags. The Tags will be included in every metric emitted by the VU and the Tags' state is maintained across different iterations until the VU exists. |

Reviewer 2: ```suggestion
| tags                | object  | The map that gives control over VU's Tags. The Tags will be included in every metric emitted by the VU and the Tags' state is maintained across different iterations while the VU exists. |
```

> | tags                | object  | The map that gives control over VU's Tags. The Tags will be included in every metric emitted by the VU and the Tags' state is maintained across different iterations while the VU exists. |

Reviewer 2: hm ... doesn't this reset between scenarios though :thinking:

Reviewer 3: Same here, I thought it was gonna reset as the cookies do.

Reviewer 2: it seems so from thsi code https://github.com/grafana/k6/blob/master/js/runner.go#L606-L623

> | tags                | object  | The map that gives control over VU's Tags. The Tags will be included in every metric emitted by the VU and the Tags' state is maintained across different iterations of the same Scenario while the VU exists. |

Reviewer 3: Suggestions: 

1 - Add a link to the example below.

2 - Add a link to k6 Tags for further reference. It could be something like: 

Read more about [k6 Tags](https://k6.io/docs/using-k6/tags-and-groups/#tags) and check out the [example](#the-links-to-example).

> > ℹ️ Setting tags

Reviewer 3: Suggestions:

1 - Rename the title to **Setting `vu.tags`**

2 - Use the [CollapsibleComponent](https://github.com/grafana/k6-docs/blob/main/CONTRIBUTING_FILE_FORMAT.md#collapsible) instead of blockquote

>

> > Set a Tag with the same key as a [SystemTag](docs/using-k6/options#system-tags) is allowed but it requires attention to avoid unexpected results. Typically, an enabled SystemTag will overwrite any duplicate of the same key that has been set before, except for the `name` SystemTag, which doesn't overwrite the same key already set for supporting the [URL Grouping](https://k6.io/docs/using-k6/http-requests/#url-grouping) feature.

Reviewer 2: ```suggestion
> Set a Tag with the same key as a [SystemTag](docs/using-k6/options#system-tags) is allowed but it requires attention to avoid unexpected results. Overwriting SystemTags will not throw an error, but in most cases will not actually change the value of the emitted metrics as expected. For example, overwriting `url` will not result in a changed value when `http.get` is called as it will set its own value, but it will add the tag `url` to the metric emitted by a `check()` or `metric.add()`. This is not true for the `name` tag and `http.*` methods as that is already supported for the purposes of the [URL Grouping](https://k6.io/docs/using-k6/http-requests/#url-grouping) feature.
```

>

> > Not all the types are accepted for a Set operation, it supports String, Number and Boolean types. The `tags` object underhood handles a Tag as a String pair of key and value, so all the types will be implicitly converted into a String. In the case, one of the denied types is used (e.g. Object or Array) and the `--throw` option is set then an exception is raised otherwise just a warning is printed and the Set operation discarded.

Reviewer: ```suggestion
> Not all the types are accepted for a Set operation: it supports String, Number and Boolean types. Under the hood, the `tags` object handles a Tag as a String pair of key and value, so all the types will be implicitly converted into a String. If one of the denied types is used (e.g. Object or Array) and the `--throw` option is set then an exception is raised otherwise just a warning is printed and the Set operation is discarded.
```
I _think_ it's more correct like this?

> > Not all the types are accepted for a Set operation: it supports String, Number and Boolean types. Under the hood, the `tags` object handles a Tag as a String pair of key and value, so all the types will be implicitly converted into a String. If one of the denied types is used (e.g. Object or Array) and the `--throw` option is set then an exception is raised otherwise just a warning is printed and the Set operation is discarded.

Reviewer 2: ```suggestion
> Not all the types are accepted for a Set operation: it supports String, Number and Boolean types. Under the hood, the `tags` object handles a Tag as a String pair of key and value, so all the types will be implicitly converted into a String. If one of the denied types is used (e.g. Object or Array) and the `--throw` option is set then an exception is thrown otherwise just a warning is printed and the Set operation is discarded.
```

> > ℹ️ Set Tag
> >
> > Set a Tag with the same key as a [SystemTag](docs/using-k6/options#system-tags) is allowed but it requires attention to avoid unexpected results. Overwriting SystemTags will not throw an error, but in most cases will not actually change the value of the emitted metrics as expected. For example, overwriting `url` will not result in a changed value when `http.get` is called as it will set its own value, but it will add the tag `url` to the metric emitted by a `check()` or `metric.add()`. This is not true for the `name` tag and `http.*` methods as that is already supported for the purposes of the [URL Grouping](https://k6.io/docs/using-k6/http-requests/#url-grouping) feature.
> >
> > Not all the types are accepted for a Set operation: it supports String, Number and Boolean types. Under the hood, the `tags` object handles a Tag as a String pair of key and value, so all the types will be implicitly converted into a String. If one of the denied types is used (e.g. Object or Array) and the `--throw` option is set then an exception is thrown otherwise just a warning is printed and the Set operation is discarded.

Reviewer 4: ```suggestion
> ℹ️ Setting tags
>
> Seting a Tag with the same key as a [system tag](docs/using-k6/options#system-tags) is allowed, but it requires attention to avoid unexpected results. Overwriting system tags will not throw an error, but in most cases will not actually change the value of the emitted metrics as expected. For example, trying to set the `url` tag value will not result in a changed tag value when `http.get()` is called, since the tag value is determined by the HTTP request itself. However, it will add the tag `url` to the metric samples emitted by a `check()` or `metric.add()`, which is probably not the desired behavior. On the other hand, setting the `name` tag will work as expected, since that was already supported for `http.*` methods, for the purposes of the [URL Grouping](https://k6.io/docs/using-k6/http-requests/#url-grouping) feature.
>
> Not all the types are accepted as a tag value: k6 supports strings, numbers and boolean types. Under the hood, the `tags` object handles a Tag as a `String` key-value pair, so all the types will be implicitly converted into a string. If one of the denied types is used (e.g. Object or Array) and the [`throw` option](https://k6.io/docs/using-k6/options/#throw) is set, an exception will be thrown. Otherwise, only a warning is printed and the tag value will be discarded.
```

```

</CodeGroup>

### Tags

> The `vu.tags` property can be used for getting or setting VU's tags.

Reviewer 3: I'd also link here to  k6 Tags to provide more context.

<CodeGroup labels={["tags-control.js"]} lineNumbers={[true]}>

```javascript
import exec from 'k6/execution';

export default function () {
  exec.vu.tags['mytag'] = 'value1';
  exec.vu.tags['mytag2'] = 2;

>   console.log(exec.vu.tags['mytag']) // "value1"
>   console.log(exec.vu.tags['mytag2']) // "2"

Reviewer 4: ```suggestion
    // the metrics these HTTP requests emit will get tagged with `mytag` and `mytag2`:
    http.batch(['https://test.k6.io', 'https://test-api.k6.io']);
```

>     // the metrics these HTTP requests emit will get tagged with `mytag` and `mytag2`:
>     http.batch(['https://test.k6.io', 'https://test-api.k6.io']);

Reviewer 4: ah, my bad, the indentation was only 2 spaces :man_facepalming: 
```suggestion
  // the metrics these HTTP requests emit will get tagged with `mytag` and `mytag2`:
  http.batch(['https://test.k6.io', 'https://test-api.k6.io']);
```
